File: algorithms/2_medium/M/01695_maximum_erasure_value/two_pointers_hashset.py
class Solution:
    def maximumUniqueSubarray(self, nums: List[int]) -> int:
        hashset = set()
        max_sum = 0
        curr_sum = 0
        left = 0
        n = len(nums)

        for right in range(n):
            num = nums[right]
            while num in hashset:
                hashset.remove(nums[left])
                curr_sum -= nums[left]
                left += 1

            curr_sum += num
            hashset.add(num)
            max_sum = max(max_sum, curr_sum)

        return max_sum

File: algorithms/2_medium/M/01695_maximum_erasure_value/two_pointers_hashset.rs
use std::collections::HashSet;

impl Solution {
    pub fn maximum_unique_subarray(nums: Vec<i32>) -> i32 {
        let mut hashset: HashSet<i32> = HashSet::new();
        let mut max_sum = 0_i32;
        let mut curr_sum = 0_i32;
        let mut left = 0_usize;
        let len = nums.len();

        for right in 0..len {
            let num = nums[right];

            while hashset.contains(&num) {
                hashset.remove(&nums[left]);
                curr_sum -= nums[left];
                left += 1;
            }

            curr_sum += num;
            hashset.insert(num);
            max_sum = max_sum.max(curr_sum);
        }

        max_sum
    }
}
